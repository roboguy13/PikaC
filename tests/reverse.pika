%generate revcat
%generate reverse

Sll : layout[-x](List);
Sll Nil := emp;
Sll (Cons head tail) :=
  exists +nxt.
  x :-> head **
  (x+1) :-> nxt **
  Sll tail [nxt];

revcat : Sll -> Sll -> Sll;
revcat Nil Nil := Nil [Sll];
revcat Nil (Cons h t) := Cons h t [Sll];
revcat (Cons h t) Nil := revcat t (Cons h (Nil [Sll]) [Sll]);
revcat (Cons h1 t1) (Cons h2 t2) := revcat t1 (Cons h1 (Cons h2 t2 [Sll]) [Sll]);

reverse : Sll -> Sll;
reverse Nil := Nil [Sll];
reverse (Cons h t) := revcat (Cons h t [Sll]) (Nil [Sll]);

%test "revcat" Sll:
  revcat (Cons 1 (Cons 2 (Nil [Sll]) [Sll]) [Sll])
    (Cons 10 (Cons 20 (Nil [Sll]) [Sll]) [Sll]);

%test "reverse" Sll:
  reverse (Cons 1 (Cons 2 (Cons 3 (Cons 4 (Nil [Sll]) [Sll]) [Sll]) [Sll]) [Sll]);
