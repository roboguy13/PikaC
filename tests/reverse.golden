*** Running test revcat
(2 (1 (10 (20 ()))))

*** Running test reverse
(4 (3 (2 (1 ()))))

